components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java

```diff
@@ -124,389 +130,410 @@
     private static final String formPostRedirectPage = getFormPostRedirectPage();
     private static final String DISPLAY_NAME = "DisplayName";
 
-    private static String getFormPostRedirectPage() {
+    @GET
+    @Path("/")
+    @Consumes("application/x-www-form-urlencoded")
+    @Produces("text/html")
+    public Response authorize(@Context HttpServletRequest request, @Context HttpServletResponse response)
+            throws URISyntaxException, InvalidRequestParentException {
 
-        java.nio.file.Path path = Paths.get(CarbonUtils.getCarbonHome(), "repository", "resources",
-                "identity", "pages", "oauth_response.html");
-        if (Files.exists(path)) {
-            try {
-                return new Scanner(Files.newInputStream(path), "UTF-8").useDelimiter("\\A").next();
-            } catch (IOException e) {
-                if (log.isDebugEnabled()) {
-                    log.debug("Failed to find OAuth From post response page in : " + path.toString());
-                }
+        startSuperTenantFlow();
+        OAuthMessage oAuthMessage = buildOAuthMessage(request, response);
+        try {
+            if (isPassthroughToFramework(oAuthMessage)) {
+                return handleAuthFlowThroughFramework(oAuthMessage);
+            } else if (isInitialRequestFromClient(oAuthMessage)) {
+                return handleInitialAuthorizationRequest(oAuthMessage);
+            } else if (isAuthenticationResponseFromFramework(oAuthMessage)) {
+                return handleAuthenticationResponse(oAuthMessage);
+            } else if (isConsentResponseFromUser(oAuthMessage)) {
+                return handleResponseFromConsent(oAuthMessage);
+            } else {
+                return handleInvalidRequest();
             }
+        } catch (OAuthProblemException e) {
+            return handleOAuthProblemException(oAuthMessage, e);
+        } catch (OAuthSystemException e) {
+            return handleOAuthSystemException(oAuthMessage.getSessionDataCacheEntry(), e);
+        } finally {
+            handleRetainCache(oAuthMessage);
+            PrivilegedCarbonContext.endTenantFlow();
         }
-        return null;
     }
 
-    @GET
+
+    @POST
     @Path("/")
     @Consumes("application/x-www-form-urlencoded")
     @Produces("text/html")
-    public Response authorize(@Context HttpServletRequest request, @Context HttpServletResponse response)
-            throws URISyntaxException {
+    public Response authorizePost(@Context HttpServletRequest request, @Context HttpServletResponse response, MultivaluedMap paramMap)
+            throws URISyntaxException, InvalidRequestParentException {
 
-        // Setting super-tenant carbon context
+        // Validate repeated parameters
+        if (!EndpointUtil.validateParams(request, paramMap)) {
```

Comment:
Why do we do the validation here?
Why not simply build the request wrapper and hand over to authorize method?

Reply:
done